web-client/public/js/update-app.js

```diff
@@ -613,61 +645,117 @@ export const updateApp = grnState => {
+        if ($(NODE_COLORING_TOGGLE_SIDEBAR).prop("checked")) {
+            if (grnState.network.expression[grnState.nodeColoring.topDataset] === undefined) {
+                let queryURLTop = buildURL({dataset: grnState.nodeColoring.topDataset});
+
+                responseData("", queryURLTop).then(function (response) {
+                    grnState.network.expression[grnState.nodeColoring.topDataset] = response;
+                    grnState.nodeColoring.nodeColoringEnabled = true;
+                    $(LOG_FOLD_CHANGE_MAX_VALUE_CLASS).removeClass("hidden");
+                    $(LOG_FOLD_CHANGE_MAX_VALUE_SIDEBAR_BUTTON).removeClass("hidden");
+                    $(LOG_FOLD_CHANGE_MAX_VALUE_HEADER).removeClass("hidden");
+
+                    if (grnState.nodeColoring.bottomDataSameAsTop) {
                         stopLoadingIcon();
-                        resolve(expressionData);
-                        updateApp(grnState);
-                    }).error(console.log("Error in accessing expression database. Result may just be loading."));
-            });
+                        updaters.renderNodeColoring();
+                    }
+                }).catch(function (error) {
+                    console.log(error.stack);
+                    console.log(error.name);
+                    console.log(error.message);
+                });
+            } else if (!grnState.nodeColoring.bottomDataSameAsTop &&
+            grnState.network.expression[grnState.nodeColoring.bottomDataset] === undefined) {
+                let queryURLBottom = buildURL({dataset: grnState.nodeColoring.bottomDataset});
+                responseData("", queryURLBottom).then(function (response) {
+                    grnState.network.expression[grnState.nodeColoring.bottomDataset] = response;
+                    grnState.nodeColoring.nodeColoringEnabled = true;
+                    $(LOG_FOLD_CHANGE_MAX_VALUE_CLASS).removeClass("hidden");
+                    $(LOG_FOLD_CHANGE_MAX_VALUE_SIDEBAR_BUTTON).removeClass("hidden");
+                    $(LOG_FOLD_CHANGE_MAX_VALUE_HEADER).removeClass("hidden");
+
+                    stopLoadingIcon();
+
+                    updaters.renderNodeColoring();
+                }).catch(function (error) {
+                    console.log(error.stack);
+                    console.log(error.name);
+                    console.log(error.message);
+                });
+            } else {
+                // Code makes it here, but node coloring fails to occur. Why???
+                grnState.nodeColoring.nodeColoringEnabled = true;
+                updaters.renderNodeColoring();
+                grnState.nodeColoring.nodeColoringEnabled = true;
+                $(LOG_FOLD_CHANGE_MAX_VALUE_CLASS).removeClass("hidden");
+                $(LOG_FOLD_CHANGE_MAX_VALUE_SIDEBAR_BUTTON).removeClass("hidden");
+                $(LOG_FOLD_CHANGE_MAX_VALUE_HEADER).removeClass("hidden");
+            }
+        }
```

Comment:
Hmmm this sequence is getting a little long, and I see a few repeated patterns. Might be good to consolidate some sequences into functions then call them when needed.